README.md

```diff
@@ -1,106 +1,181 @@
 
 ### Non-goals
 
-[If there are "adjacent" goals which may appear to be in scope but aren't, enumerate them here. This section may be fleshed out as your design progresses and you encounter necessary technical and other trade-offs.]
+- Cover all application requirements for a key-value store. Instead, it focuses on the most common use cases. It allows providers to extend the API to support more use cases.
+- Data consistency.
+- Data replication.
+- Data sharding.
+- Cluster management.
+- Monitoring
+
 
-### API walk-through
 
-[Walk through of how someone would use this API.]
+
+### API walk-through
 
 #### [Use case 1]
 
-[Provide example code snippets and diagrams explaining how the API would be used to solve the given problem]
+Imagine you have a HTTP handler that needs to persistent some data to a persistent key-value store. The handler needs to be able to retrieve the data from the key-value store, and update the data in the key-value store. The handler also needs to be able to delete the data from the key-value store. The following rust code shows how you can use the WASI Key-Value Store API to in the handler.
 
-#### [Use case 2]
+```rust
+let value = b"spiderlightning";
+kv1::put("key", value)?;
+let val = kv1::get("key")?;
+assert_eq!(val, value);
 
-[etc.]
+kv1::delete("key")?;
+```
 
 ### Detailed design discussion
 
-[This section should mostly refer to the .wit.md file that specifies the API. This section is for any discussion of the choices made in the API which don't make sense to document in the spec file itself.]
+```go
+interface "wasi:kv/data/crud" {
+  use { Error, payload, payload_stream } from "wasi:kv/types"
 
-#### [Tricky design choice #1]
+	get: func(key: string) -> result<payload_stream, Error>
 
-[Talk through the tradeoffs in coming to the specific design point you want to make.]
+  put: func(key: string, value: payload_stream) -> result<_, Error>
 
+  del: func(key: string) -> result<_, Error>
+}
+
+interface "wasi:kv/data/increment" {
+  use { Error, payload, payload_stream } from "wasi:kv/types"
+  
+	incr: func(key: string) -> result<u64, Error>
+}
+
+interface "wasi:kv/data/bulk-get" {
+  use { Error, payload, payload_stream } from "wasi:kv/types"
+  
+	bulk-get: func(keys: stream<string>) -> result<stream<(key, payload_stream)>, Error>
+
+  keys: func() -> result<future<list<string>>, Error>
+}
+
+interface "wasi:kv/data/bulk-put" {
+  use { Error, payload, payload_stream } from "wasi:kv/types"
+  
+	bulk-put: func(key_values: stream<(string, payload)>) -> result<unit, Error>
+}
+
+interface "wasi:kv/data/ttl" {
+  use { Error, payload, payload_stream } from "wasi:kv/types"
+  
+	put-with-ttl: func(key: string, value: payload, ttl: u64) -> result<unit, Error>
+}
+
+interface "wasi:kv/data/query" { 
+  ...
+}
+
+
+interface "wasi:kv/types" {
+	resource Error { ... }
+
+  type payload_stream = stream<u8>
+	type payload = list<u8>
+}
```

Comment:
For record-keeping sake, I'm copy-pasting the feedback I gave prior.

> - Why make the get function return a `payload_stream` rather than just a `future<payload>`?
> - I noticed you changed set to `put` in the crud interface, but not for the `bulk-*` interfaces. Is that an oversight? > I'd probably just recommend keeping it as `set`...
> - For `bulk-get`, do we really want a list of streams or should one stream be enough (perhaps streaming a list of payloads...?)?
> - Why separate the two `bulk-*` functions into two interfaces and not just make a single `bulk` interface?

~most of these have been addressed already ✔️

Reply:
Copy-and-pasted my comments on this question:

- There are some document stores, such as MongoDB and DynamoDB, can store a JSON file for a key as value. This means that the value from get could be pretty big. It makes more sense to get a stream of u8 instead of waiting for a future of payload to come. If you take a look at Luke's presentation on HTTP Body type, it is also defined as a stream. The biggest difference between stream and future is that stream is list of futures.

- For clarity, I changed the name to `create_or_update`. I am very flexible on the naming of the APIs. Let me know if you have any other suggestions!

- Done

- I've seen some datastores only support bulk-get, but not bulk-set